File: Awful.apk/src/main/java/com/ferg/awfulapp/forums/DropdownParserTask.java
package com.ferg.awfulapp.forums;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;
import android.widget.Toast;

import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.util.AwfulError;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * <p>Created by baka kaba on 18/04/2016.</p>
 * <p/>
 * <p>Forum update task that parses the forum jump dropdown.
 * This task only requires a single page load, but can't scrape subtitles
 * and misses certain forums.</p>
 * <p/>
 * <p>It works by finding the set of <code>option</code> tags in the dropdown
 * box, then looks for the items that are formatted like forum entries.
 * It keeps a running model of the current branch hierarchy, so it can identify
 * what level a forum should be on and which forum is its parent</p>
 */
public class DropdownParserTask extends UpdateTask {

    private final List<Forum> parsedForums = new ArrayList<>();

    {
        TAG = "DropdownParserTask";
    }


    public DropdownParserTask(@NonNull Context context) {
        super(context);
        initialTask = new DropdownParseRequest();
    }


    private class DropdownParseRequest extends ForumParseTask {

        {
            url = Constants.FUNCTION_FORUM + "?" + Constants.PARAM_FORUM_ID + "=" + Constants.FORUM_ID_GOLDMINE;
        }

        @Override
        protected void onRequestSucceeded(Document doc) {
            Log.i(TAG, "Got page - parsing dropdown to get forum hierarchy");
            parsePage(doc);
        }


        @Override
        protected void onRequestFailed(AwfulError error) {
            Log.w(TAG, "Request error - couldn't get page to parse");
        }
    }


    private void parsePage(Document doc) {
        // can't do anything without a page - fail immediately
        if (doc == null) {
            fail("no document to parse");
            return;
        }

        class ParsedForum {
            final int id;
            final int depth;


            ParsedForum(int id, int depth) {
                this.id = id;
                this.depth = depth;
            }
        }

        try {
            // this stack works like a breadcrumb trail, so we can compare to the last forum
            // added at the current depth, and work out the new forum's relationship to it
            Stack<ParsedForum> parsedForumStack = new Stack<>();
            // this is a dummy forum representing the root - every other forum will be below this (depth >= 0)
            parsedForumStack.push(new ParsedForum(ForumRepository.TOP_LEVEL_PARENT_ID, -1));
            ParsedForum previousForum;

            // get the items from the dropdown
            Elements forums = doc.select("form.forum_jump option");
            Log.d(TAG, "Found " + forums.size() + " elements");
            for (Element option : forums) {
                String idString = option.val();
                String rawTitle;
                try {
                    // we need the raw text so we can preserve the leading space that identifies sections like Main
                    rawTitle = option.textNodes().get(0).getWholeText();
                } catch (IndexOutOfBoundsException e) {
                    throw new ParseException("Couldn't get TextNode for title", 0);
                }

                Integer forumDepth = getForumDepth(rawTitle);
                if (forumDepth == null) {
                    // not a forum
                    continue;
                }
                // strip off the depth-marking prefix (dashes plus a leading space)
                String title = rawTitle.substring(forumDepth + 1);

                // if this new forum isn't as deep as the previous one on the stack, it's a new branch
                // and we need to back up. Pop them off until we find a forum on the same or higher level
                previousForum = parsedForumStack.peek();
                while (previousForum.depth > forumDepth) {
                    parsedForumStack.pop();
                    previousForum = parsedForumStack.peek();
                }

                if (forumDepth == previousForum.depth) {
                    // the new forum is a sibling of the one on the stack - remove that so
                    // this will take its place as the last added on this level (and the one
                    // on the stack is its parent)
                    parsedForumStack.pop();
                    previousForum = parsedForumStack.peek();
                }


                int id;
                try {
                    id = Integer.parseInt(idString);
                } catch (NumberFormatException e) {
                    throw new ParseException("Can't parse forum ID as int! Got: " + idString, 0);
                }
                // we're going to push the new forum onto the stack - the one currently on top will be its parent
                int parentId = previousForum.id;
                parsedForums.add(new Forum(id, parentId, title, ""));
                parsedForumStack.push(new ParsedForum(id, forumDepth));

                if (Constants.DEBUG) {
                    Log.d(TAG, String.format("(ID:%s Parent:%d) %s", id, parentId, title));
                }
            }
        } catch (ParseException e) {
            Log.e(TAG, "Unexpected parse failure - has the page formatting changed?", e);
            if (Constants.DEBUG) {
                Toast.makeText(context, "DROPDOWN PARSE ERROR\n" + e.getMessage(), Toast.LENGTH_LONG).show();
            }
        }
    }


    /**
     * Get the depth of a forum by its title prefix in the dropdown.
     * The value is the number of leading dash characters, so the actual
     * depth is relative (e.g. one level below -- is ----, so one level deeper than 2 is 4)
     *
     * @param rawTitleText The raw, unnormalised text from the dropdown (the whitespace is important)
     * @return The depth value, which is the number of leading dashes (so you can strip them for display)
     */
    @Nullable
    private Integer getForumDepth(@NonNull String rawTitleText) {
        if (rawTitleText.isEmpty()) {
            return null;
        }

        // Valid forum titles have some number of - chars followed by a space, so get the number of -s
        for (int i = 0; i < rawTitleText.length(); i++) {
            char c = rawTitleText.charAt(i);
            if (c == ' ') {
                return i;
            } else if (c != '-') {
                // if we run into a non-dash before we find a space, give up
                break;
            }
        }

        // we didn't find a space after some dashes, so this isn't a valid forum title!
        return null;
    }


    @NonNull
    @Override
    protected ForumStructure buildForumStructure() {
        return ForumStructure.buildFromOrderedList(parsedForums, ForumRepository.TOP_LEVEL_PARENT_ID);
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/forums/DropdownParserTask.kt
package com.ferg.awfulapp.forums

import android.content.Context
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.util.AwfulError
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import timber.log.Timber
import java.text.ParseException
import java.util.*

/**
 *
 * Created by baka kaba on 18/04/2016.
 *
 * Forum update task that parses the forum jump dropdown.
 * This task only requires a single page load, but can't scrape subtitles
 * and misses certain forums.
 *
 * It works by finding the set of `option` tags in the dropdown
 * box, then looks for the items that are formatted like forum entries.
 * It keeps a running model of the current branch hierarchy, so it can identify
 * what level a forum should be on and which forum is its parent
 */
internal class DropdownParserTask(context: Context) : UpdateTask(context) {

    override val initialTask: ForumParseTask = DropdownParseRequest()
    private val parsedForums = ArrayList<Forum>()


    private inner class DropdownParseRequest : UpdateTask.ForumParseTask() {

        override val url: String
            get() = "$FUNCTION_FORUM?$PARAM_FORUM_ID=$FORUM_ID_GOLDMINE"

        override fun onRequestSucceeded(doc: Document) {
            Timber.i("Got page - parsing dropdown to get forum hierarchy")
            parsePage(doc)
        }

        override fun onRequestFailed(error: AwfulError) {
            Timber.w("Request error - couldn't get page to parse")
        }
    }


    /**
     * Parse a forum hierarchy from a page containing a dropdown picker, populating [parsedForums]
     */
    private fun parsePage(doc: Document?) {
        // can't do anything without a page - fail immediately
        if (doc == null) {
            fail("no document to parse")
            return
        }

        try {
            // this stack works like a breadcrumb trail, so we can compare to the last forum
            // added at the current depth, and work out the new forum's relationship to it
            val forumHierarchy = Stack<ParsedForum>().apply {
                // this is a dummy forum representing the root - every other forum will be below this (depth >= 0)
                push(ParsedForum(ForumRepository.TOP_LEVEL_PARENT_ID, "root", -1))
            }

            fun getPreviousForum(): ParsedForum = forumHierarchy.peek()

            // get the items from the dropdown
            val items = doc.select("form.forum_jump option")
            Timber.d("Found ${items.size} elements")
            if (items.isEmpty()) throw ParseException("No dropdown items found!", 0)

            items.mapNotNull(::parseForum).forEach { forum ->
                // backtrack until we get to a forum at a higher level than this forum, so we can insert it below
                while (getPreviousForum().depth >= forum.depth) {
                    forumHierarchy.pop()
                }

                // we're going to push the new forum onto the stack - the one currently on top will be its parent
                val parentId = getPreviousForum().id
                parsedForums.add(Forum(forum.id, parentId, forum.title, ""))
                forumHierarchy.push(forum)
                if (DEBUG) Timber.d(" ${forum.title} (ID:${forum.id} Parent:$parentId)")
            }
        } catch (e: ParseException) {
            Timber.e(e, "Unexpected parse failure - has the page formatting changed?")
        }

    }


    // Regex for identifying a forum title and extracting its indent(if any) and title content
    // format is [optional indent dashes] [required space] [required title text]
    private val forumTitleRegex = Regex("""^(-*) +(.+)""")

    /**
     * Parse forum details from an item in the dropdown picker.
     *
     * This will return null for non-standard forum items (e.g. Private Messages) and things like
     * dividers. Depth is signified by the number of dashes preceding the title - right now this is
     * 2 per level (so "---- Forum 2" is below "-- Forum 1") but this could change, so we just use the
     * raw value and treat the numbers as relative (if it's equal it's the same level, if it's bigger
     * it's a lower level, etc)
     */
    @Throws(ParseException::class)
    private fun parseForum(dropdownItem: Element): ParsedForum? {
        // we need the raw text so we can preserve the leading space that identifies sections like Main
        val rawTitle = dropdownItem.textNodes().firstOrNull()?.wholeText
                ?: throw ParseException("Couldn't get TextNode for title", 0)
        // we're expecting leading dashes followed by a space, and then some characters
        val result = forumTitleRegex.find(rawTitle)
                ?: return null
        // get the forum ID - this will fail for non-forums ("Private Messages" etc) as they have non-int IDs,
        // which is why we do this after verifying it is a forum
        val idString = dropdownItem.`val`()
        val forumId = idString.toIntOrNull()
                ?: throw ParseException("Can't parse forum ID as int! Got: $idString", 0)
        return result.destructured.let { (dashes, title) -> ParsedForum(forumId, title, depth = dashes.length) }
    }


    override fun buildForumStructure(): ForumStructure =
            ForumStructure.buildFromOrderedList(parsedForums, ForumRepository.TOP_LEVEL_PARENT_ID)

    private data class ParsedForum(val id: Int, val title: String, val depth: Int)
}


